[로또 1단계] 모찌 미션 제출합니다.

wondroid-world opened the pull request:
안녕하세요. 두루!
저는 고양이 두마리(모모, 치치)를 키우고 있는 모찌입니다. 만나서 반갑습니다.


# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?


# 어떤 부분에 집중하여 리뷰해야 할까요?

### **설계 방향성**
- LottoNumber(로또 번호 각각을 위한 클래스) -> Lotto(로또 한줄을 위한 클래스)
- LottoMachine(로또 발행 기계)
- Scanner -> Result(로또 결과를 판단하는 클래스)



저희는 MVC 중 **model의 중요성이 높다**고 생각했습니다. 


그래서 model에서 주요 데이터를 다루고 데이터를 수정하는 메서드를 생성했습니다. 


controller에서는 model과 view의 연결 다리 역할만 했으면 좋겠다고 생각했습니다. controller는 가볍게 만들기를 지향했습니다. 


view에서는 값을 받고 기능요구 사항에 영향이 크지 않는 부분만 빼서 검증을 진행했습니다. 


## 코드 리뷰 커뮤니케이션

###**질문**
1. **상수화를 어디까지 해야할까요?**

```
 fun printResult(results: Map<Rank, Int>) {
        println("\n당첨 통계")
        println("---------")
        for (result in results) {
            val matchMessage =
                when (result.key) {
                    Rank.FIRST -> "6개 일치"
                    Rank.SECOND -> "5개 일치, 보너스 볼 일치"
                    Rank.THIRD -> "5개 일치"
                    Rank.FOURTH -> "4개 일치"
                    Rank.FIFTH -> "3개 일치"
                    Rank.MISS -> "0개 일치"
                }
            val prizeMessage =
                when (result.key) {
                    Rank.FIRST -> "2000000000원"
                    Rank.SECOND -> "30000000원"
                    Rank.THIRD -> "1500000원"
                    Rank.FOURTH -> "50000원"
                    Rank.FIFTH -> "5000원"
                    Rank.MISS -> "0원"
                }

            println("$matchMessage ($prizeMessage) - ${result.value}개")
        }
    }
```

페어랑 이야기한 부분에서 이 outputview에서 나오는 값 모두 상수화를 해야하나 고민을 했습니다. 하지만, 제가 내린 결론은 유지보수 관점에서 상수화하는 것보다 값을 그대로 사용하는 것이 좋다고 생각했습니다. 변수를 아무리 잘 지어도, 각 줄을 읽을 때마다 상수화된 값을 찾아 읽어야하는 번거로움이 있기때문입니다. 


두루는 저의 생각에 대해서 어떻게 생각하시나요?


2. **현업 관점에서 로또 당첨 번호 클래스와 로또 자동 번호 클래스는 분리해야하나요?**


저는 로또의 당첨을 판단하는 로직은 Scanner와 Result에 있습니다. 


로또 당첨 번호는 그 번호만 잘 보관하고 있으면 되므로 로또 자동 생성 번호와 똑같다고 생각했습니다. (보너스 번호 제외) 그래서 똑같은 class를 사용하고 있습니다.  저는 재사용 측면에서 좋다고 생각하는 데, 위험한 걸까요?


3. validation은 어디서 해야할까?


두루 저는 이번에 기능 요구 사항에서 중요한 부분은 model 안에서 검사를하고 input이 빈값이거나 중요하지 않은 로직은 input에서 진행을 했습니다.


제 생각에는 유지보수 관점에서는 validation class를 만들어서 validate를 다 이임하는 것이 좋다고 생각합니다. 왜냐하면 한곳에 로직이 모여 있으니, 찾고 수정하기 편하다라는 일차원적인 생각 때문입니다.


객체 지향적인 관점에서 각자가 하는 일을 믿고 위임해줘야한다고 생각합니다. 


하지만, 이번에는 페어의 의견에 따라서 분리를 해봤습니다. 지금 프로그램처럼 하면, 응집도가 높아 좋은거 같은 거 같고 너무 고민이네용....

wondroid-world committed:
docs: 기능 목록 작성

wondroid-world committed:
docs: 기능 목록 수정

wondroid-world committed:
feat: 로또 번호 뽑기

wondroid-world committed:
feat: 리스트에 숫자 추가

wondroid-world committed:
feat: LottoNumbers 클래스 구현

wondroid-world committed:
feat: 로또 구입 개수 만큼 로또 발행

wondroid-world committed:
feat: 로또 머신 기능 구현

wondroid-world committed:
feat: 로또 번호가 몇개 일치하는 지 확인

wondroid-world committed:
feat: 수익률 계산

HamBeomJoon committed:
docs: 기능 목록 수정

HamBeomJoon committed:
feat: 로또 번호 생성

HamBeomJoon committed:
feat: 로또 클래스 생성

HamBeomJoon committed:
feat: 로또 구매 개수만큼 로또 발행

HamBeomJoon committed:
feat: 로또 당첨 등수 확인

HamBeomJoon committed:
feat: 로또 당첨 결과 저장

HamBeomJoon committed:
feat: 수익률을 계산

HamBeomJoon committed:
feat: 입력 함수 구현, 유효성 검사

HamBeomJoon committed:
feat: 보너스 번호 중복 예외 처리

HamBeomJoon committed:
refactor: 모델 패키지 변경

HamBeomJoon committed:
fix: 구입 금액 나머지 검사 수정

HamBeomJoon committed:
feat: 출력 함수 구현

HamBeomJoon committed:
feat: lottoStore 로직 구현

HamBeomJoon committed:
feat: application 실행 함수

HamBeomJoon committed:
feat: 당첨 결과, 수익률 출력

HamBeomJoon committed:
refactor: 로또 번호 유효성 검사 상수화

HamBeomJoon committed:
refactor: 로또 구입 금액 유효성 검사 상수화

HamBeomJoon committed:
feat: 구입 금액 음수 예외처리

HamBeomJoon committed:
refactor: 상수화

HamBeomJoon committed:
chore: 파일 정리

HamBeomJoon committed:
chore: 패키지 이동

HamBeomJoon committed:
test: 지난 주 당첨 번호 테스트

HamBeomJoon committed:
refactor: 함수 접근 제어자 수정

HamBeomJoon committed:
refactor: 파일 정리

HamBeomJoon committed:
refactor: 테스트 패키지 수정

HamBeomJoon committed:
refactor: 함수 접근 제어자 수정

HamBeomJoon committed:
refactor: run 함수 분리

Gyuil-Hwnag commented on README.md:
```diff
@@ -1 +1,28 @@
-# kotlin-lotto
\ No newline at end of file
+# kotlin-lotto
+
+## 기능 요구 사항
```
README 작성 정말 좋습니다~👍

wondroid-world replied:
칭찬 감사합니다. 두루*

Gyuil-Hwnag commented on src/main/kotlin/lotto/Application.kt:
```diff
@@ -0,0 +1,6 @@
+package lotto
+
+fun main() {
+    val lottoStore = LottoStore()
+    lottoStore.run()
+}
```
추후 게임종류가 많아진다면, main() 이라는 네이밍이 다소 애매하게 느껴질 수 있겠네요~
이를 로또에 맞게 바꾸려면 어떻게 바꿔보는 것이 좋을까요?🤔

wondroid-world replied:
 두루! 
 제가 알기로, JVM의 컴파일러가 실행 진입점으로 찾을려면 main으로 되어있다고 알고있습니다! Java의 public static void main(String arg[])가 시작점이 되듯이 코틀린은 fun main으로 되어있어야 한다고 알고 있습니다!
 
혹시 제가 이해를 잘 못한 걸까요? 이야기 주신 부분이 Application.kt보다 Lotto에 맞는 파일명으로 바꿔보는 것을 이야기 주는걸까요? 🤔
 
 그럼, "Lotto"이라고 적어야지 모든 파일을 포옹하는 느낌인거 같아요! 두루덕분에 Application.kt 이름을 변경해도 되는 거 처음 알았습니다. 감사합니당!

wondroid-world replied:
[변경한 커밋](https://github.com/wondroid-world/kotlin-lotto/commit/a1503191bea3c69992560021c5aa66e532f4ef87)

Gyuil-Hwnag replied:
아 넵 파일명 맞습니다!!

Gyuil-Hwnag commented on src/main/kotlin/lotto/Constants.kt:
```diff
@@ -0,0 +1,7 @@
+package lotto
+
+object Constants {
+    const val LOTTO_AMOUNT = 1_000
+    const val MINIMUM_NUMBER = 1
+    const val MAXIMUM_NUMBER = 45
+}
```
상수화도 잘해주셨네요~
추후 상수 관련된 내용이 많아진다면, 모두 Constants 내에 모이게 되는걸까요?
모찌의 의견을 남겨주세요ㅎㅎ💪

wondroid-world replied:
저희가 상수화 중 프로그램 전체에서 사용되는 부분은 Constants에 넣었습니다.

확장성측면에서 Contants에 모든 상수를 넣으면 좋겠다라고 생각했습니다. "다만 현 상황에서, model 안에서 사용만하는 부분을 JVM의 메모리 중 스택 영역에 쌓아두는 것이 메모리를 덜 차지할거 같다" 라는 생각에서 model에 상수화를 했습니다.

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,31 @@
+
+class Lotto(
+    private val numbers: List<LottoNumber> = generateLotto(),
+    private val bonusNumber: LottoNumber? = null,
+) {
```
bonusNumber를 nullable 하게 설정하신 이유가 있을까요..?

wondroid-world replied:
로또 자동 생성 번호랑 로또 당첨 번호랑 같은 클래스에서 관리하고 싶었습니다. 

로또 자동 생성 번호에서는 bonusNumber가 필요없으니 null으로 설정되어야 하고 로또 당첨 번호는 bonusNumber를 가져야하니, null이 허용될 수 있는 타입으로 설정했습니다.

wondroid-world replied:
406fd17f1c42c14cc5ac32ec516344cc204aabae

하지만, 크루와 코치분들과 이야기하면서, 생각이 바뀌어서, Lotto 클래스와 WinningLotto 클래스를 분리했습니다. 

왜냐하면 로또 판단하는 로직이 여러개이면 LottoScanner에서 관리하는 것이 맞다고 생각을 합니다. 하지만, 로또 당첨 판단 로직이 하나이면 이는 WinningLotto에서 데이터와 데이터 관련 로직 모두를 가지고 있으면 좀더 캡슐화 면에서 더 옳다고 판단되어 로직을 변경했습니다.

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,31 @@
+
+    fun getSize() = numbers.size
+
+    fun getNumbers() = numbers.toList()
+
```
이미 numbers는 List Collection 인데, toList()를 붙히신 이유가 궁금합니다!

wondroid-world replied:
제가 LottoTicket을 일급 객체를 사용했습니다.
밖에서 LottoTicket 안의 번호를 읽게 하기위해서, toList()로 변경을 진행했습니다!

wondroid-world replied:
아! 맞네요! 수정했습니다. 감사합니당!

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,31 @@
+
+    fun getBonusNumber() = bonusNumber
+
+    companion object {
+        private fun generateLotto(): List<LottoNumber> =
+            (Constants.MINIMUM_NUMBER..Constants.MAXIMUM_NUMBER)
+                .shuffled()
+                .take(LOTTO_PICK_COUNT)
+                .sorted()
+                .map { LottoNumber(it) }
+
+        private const val LOTTO_PICK_COUNT = 6
+    }
```
Companion object의 활용 좋네요ㅎㅎ👍

한가지 궁금한게 있습니다!
로또 번호 생성도 로또 객체의 책임이라고 지정을 하신 이유가 궁금합니다~

wondroid-world replied:
처음 저의 생각에서는 로또가 수동으로 번호가 입력되는 순간 아니면 무조건 번호를 발급해야한다고 생각했습니다. 그래서, 로또에 역할을 부여했네요.

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/LottoMachine.kt:
```diff
@@ -0,0 +1,17 @@
+
+class LottoMachine(
+    count: Int,
+) {
+    val lottos: List<Lotto> = generateLottos(count)
```
LottoMachine.lottos로 보니까 뭔가 어떤기능을 담당하는지 잘 눈에 안들어오는 느낌이 들었는데, 모찌의 생각은 어떠신가요?
아래의 두개를 보았을 때, 로직을 모르는 사람이 본다면 어떤 것이 더 한눈에 들어올까요?🤔
~~~kotlin
val purchaseLottos = LottoMachine.purchase(count)
vs
val purchaseLottos = LottoMachine.lottos
~~~

wondroid-world replied:
두루처럼 
LottoMachine.purchase(count)
와 같이 하면, 읽는 사람이 더 편할 꺼 같네용! 
좋은 코멘트 감사합니다!

wondroid-world replied:
7453114c0eec3ad1f883d5c3376fee9273050bac

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/LottoNumber.kt:
```diff
@@ -0,0 +1,17 @@
+
+import lotto.Constants
+
+data class LottoNumber(
+    val number: Int,
+) {
+    init {
+        require(number in Constants.MINIMUM_NUMBER..Constants.MAXIMUM_NUMBER) {
+            IllegalArgumentException(ERROR_INVALID_RANGE)
+        }
+    }
+
+    companion object {
+        private const val ERROR_INVALID_RANGE = "숫자는 1에서 45 사이여야 합니다."
+    }
+}
```
LottoNumber 객체의 책임을 잘 분리해주셨네요ㅎㅎ👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/LottoScanner.kt:
```diff
@@ -0,0 +1,20 @@
+package lotto.model
+
+class LottoScanner(
+    private val winningNumbers: Lotto,
+) {
```
LottoScanner 객체의 구현 정말 잘해주셨네요~👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,31 @@
+
+import lotto.Constants
+
+class Lotto(
+    private val numbers: List<LottoNumber> = generateLotto(),
```
> 현업 관점에서 로또 당첨 번호 클래스와 로또 자동 번호 클래스는 분리해야하나요?

제가 질문을 재대로 이해를 한게 맞는지 한번 봐주세요! 
(WinnerLotto / Lotto 두개로 분리를 했어야 하는지에 대한 질문일까요?)
 
그렇다면 우선 담청 로또 번호와 로또 번호와 구분해야 겠다고 고민을 하신 이유가 궁금합니다!
두개는 어떤 차이점이 있으며, 어떠한 차이점 때문에 구분을 하려고 하신 이유가 궁금하네요..!🤔

wondroid-world replied:
당첨 로또와 자동 발행되는 로또는 똑같이 로또 번호 6개를 가집니다. 하지만, 당첨 순위를 정확하게 판단하기 위해서, 당첨 로또는 보너스 번호까지 생성자에 포함해야한다고 생각했습니다. 

저는 똑같이 로또 번호 6개를 가지고 있다는 특성 때문에, 로또 클래스에서 당첨 로또와 자동 발행되는 로또 모두다 생성하고 싶었습니다. 하지만, 지금 생각해보니, 둘의 역할이 다른거 같아서, 현재 refactoring한 코드에서는 자동 발행되는 로또와 당첨 로또 클래스를 각각 만들었습니다.

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,57 @@
+
+import lotto.Constants
+
+class InputView {
+    fun validateAmount(amount: String): Int {
```
위에 Constant 에서 남겨둔 리뷰와 동일하게 어떤 View인지에 대해서 드러나는 것도 좋을 것 같다는 생각이 듭니다~💪

wondroid-world replied:
두루! 제가 두루의 의도를 이해를 못했는 뎅..... 😢 
자세히 설명 부탁드려도 될까요?

Gyuil-Hwnag replied:
Constant 코멘트에서 제가 남겨두고 싶은 내용이 하나 빠져있었군요!
이번 코멘트에 남겨두겠습니다

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/OutputView.kt:
```diff
@@ -0,0 +1,50 @@
+        println(lotto.getNumbers().map { it.number }.joinToString(", ", "[", "]"))
+    }
+
+    fun printResult(results: Map<Rank, Int>) {
+        println("\n당첨 통계")
```
> 상수화를 어디까지 해야할까요?

개인적으로 하드코딩을 지양하고 왠만하면, 상수화를 하자! 이긴 합니다
실제로 개발을 하시다보면, 재 사용이 안될 것 같은 것들도 재사용이 되는 경우도 많더라구요..
그리고 상수화를 함으로써 얻는 장점은 한곳에서 확인이 가능하기도 합니다!
프로젝트가 커져서, 무수히 많은 코드 라인이 생긴다면 한곳에서 확인이 가능한 장점이 생기게 됩니다

그치만 오히려 한곳에서 확인이 가능하지만, 말씀해주신대로 실제 구현부분에 어떤 상수인지에 대해서 네이밍만 보고 알아보기 힘들어서 댑스가 깊어지기는 합니다

그럼에도 불구하고 실제 개발을 할때에는 왠만하면! 하드 코딩을 지양하고 있기는 합니다!
(리브랜딩이나 문구 수정이 일어날 경우, 하드코딩이 되어 있다면 여러곳을 탐방하면서 확인을 해야하는 등 여러가지 이유로 인해서요)

Gyuil-Hwnag commented on src/test/kotlin/lotto/model/LottoMachineTest.kt:
```diff
@@ -0,0 +1,14 @@
+import org.junit.jupiter.params.provider.ValueSource
+
+class LottoMachineTest {
+    @ParameterizedTest
+    @ValueSource(ints = [1, 3, 4])
```
파라미터 테스트 좋네요ㅎㅎ👍

Gyuil-Hwnag commented on src/test/kotlin/lotto/model/LottoTest.kt:
```diff
@@ -0,0 +1,29 @@
+import org.junit.jupiter.api.assertThrows
+
+class LottoTest {
+    @Test
+    fun `로또는 한 장에 로또 번호 6개를 가진다`() {
+        val lotto = Lotto()
+        assertEquals(lotto.getSize(), 6)
+    }
```
이건 한번 고민해 보면 좋을 것 같은 내용이라서 참고차 한번 남겨두었습니다~
`로또 범위를 지정하고, 생성하는 객체를 만들어 볼 수도 있지 않을까?` 그렇다면, 로또 생성에 대한 테스트도 가능할 것 같은데? 라는 생각을 해보았어요

wondroid-world replied:
두루, 맞아요! 저도 이 부분이 고민되는 부분이었습니다.  

처음 만들때에는 로또 한장(LottoTicket)이 생성될 때에는 무조건 번호가 포함되어야한다고 해서 분리하지 않았습니다. 

이 부분은 조금 더 배우고 고민하고 코드를 수정해보겠습니다!

Gyuil-Hwnag commented on src/test/kotlin/lotto/model/LottoTest.kt:
```diff
@@ -0,0 +1,29 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+class LottoTest {
+    @Test
+    fun `로또는 한 장에 로또 번호 6개를 가진다`() {
```
참고해보면 좋을 내용인 것 같아서 남겨두었습니다!

https://developer.android.com/studio/test/test-in-android-studio#view_test_coverage
테스트코드를 어느 정도로 작성했는지 확인하는 방법으로 테스트 커버리지 라는 개념이 있습니다
물론 IDEA에서 기능을 지원해 주고 있습니다~

Gyuil-Hwnag commented on src/test/kotlin/lotto/model/LottoTest.kt:
```diff
@@ -0,0 +1,29 @@
+        assertEquals(lotto.getSize(), 6)
+    }
+
+    @Test
+    fun `로또 번호는 서로 중복되지 않는다`() {
+        assertThrows<IllegalArgumentException> {
+            Lotto(
+                listOf(
+                    LottoNumber(1),
+                    LottoNumber(2),
+                    LottoNumber(3),
+                    LottoNumber(4),
+                    LottoNumber(5),
+                    LottoNumber(5),
+                ),
+            )
+        }
+    }
```
참고할 만한 내용이 있어서 코멘트 남겼습니다~
테스트 코드의 경우 어떤 테스트인지 확실하게 알 수 있게 네이밍을 정해주는 것도 좋은 것 같습니다!
온보딩 미션인 자동차 경주의 코드를 예시로 들었습니다
~~~kotlin
@Test
fun `자동차가 1칸 이동하면 거리는 1이다`() {
// given
val myCar = Car("Duru")

// when
myCar.move(1)

// result
assertThat(myCar.distance).isEqualTo(1)
}
~~~

wondroid-world replied:
```
fun `로또 번호는 서로 중복되면 에러가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            LottoTicket(
                setOf(
                    LottoNumber(1),
                    LottoNumber(2),
                    LottoNumber(3),
                    LottoNumber(4),
                    LottoNumber(5),
                    LottoNumber(5),
                ),
            )
        }
    }
```

두루!
이렇게 네이밍을 지으면 어떤 테스트인지 확실한 지 알 수 있을까요?

Gyuil-Hwnag replied:
지금도 네이밍은 좋아 보입니다!
사실 해당 코멘트의 중요한점 중 또 한가지는, `주어진 것 / 액션 / 예상되는 결과`를 함수 내에서 구분을 해보는 것 입니다~
(그래서 예시에 given / when / result 를 살포시 넣어두었었습니다)

여러 객체들이 유기적으로 연결되어 있는 테스트의 경우에는 이들이 구분이 되어 있지 않다면, 보는 사람도 헷갈리고 일정 기간이 지난후 작성했던 제가 보기에도 한눈에 안들어온다는 느낌을 받았어서요

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,57 @@
+        require(amount.toIntOrNull() != null) {
+            IllegalArgumentException(ERROR_INVALID_AMOUNT)
+        }
+        require(amount.toInt() > ZERO) {
+            IllegalArgumentException(ERROR_INVALID_OVER_ZERO)
+        }
+        require(amount.toInt() % Constants.LOTTO_AMOUNT == ZERO) {
+            IllegalArgumentException(ERROR_INVALID_DIVISION)
+        }
+        return amount.toInt()
```
> validation은 어디서 해야할까?

개인적인 생각으로는 각 객체의 책임에 맞게 Validation을 하는 것이 좋다는 생각이 들긴 합니다!
근데 Model(도메인)의 경우 각 객체에 책임에 맞게 하는 것이 좋은데, View의 경우 빈칸 판별 로직등은 Validator를 통해서 한다면 재 사용이 가능하며, 중복된 구현을 줄여볼 수 있을 것 같습니다!

그 보다 더 중요한 건 모찌만의 기준을 세우고 그에 대한 근거를 가지고 그에 맞춰서 하는게 제일 중요하다는 의견입니다~

wondroid-world replied:
좋은 의견 감사합니다!

Gyuil-Hwnag reviewed (changes requested):
모찌 안녕하세요!
이번 미션을 함께하게 된 두루라고 합니다. 잘부탁드립니다!

해당 미션 구현 고생 많으셨습니다!
미션 구현을 너무 잘 해주셨습니다 💯

고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

wondroid-world committed:
refactor: Application.kt -> Lotto.kt로 파일명 변경

- Lotto 프로그램에 맞는 파일명으로 변경

wondroid-world committed:
refeator: Lotto class -> LottoTicket class 네이밍 변경

wondroid-world committed:
refeator: Lotto class -> LottoTicket class 네이밍 변경

wondroid-world committed:
refactor: LottoScanner -> WinningLotto로 class 변경

- 로또 판별 로직은 변하지 않기 때문에 WinningLotto 클래스를 생성하면서, 판별하는 로직 생성

wondroid-world committed:
refactor: 함수명 변경

wondroid-world committed:
refactor: main함수 파일명 변경

wondroid-world committed:
refactor: 랭크 순서 변경

wondroid-world committed:
feat: WinningLotto class 생성

wondroid-world committed:
refactor: LottoTicket 역할 축소

- 과거 로또 당첨 번호와 로또 자동 발행 모두 관리하는 클래스였음
- 한 객체가 2역할을 하는 것과 같아 로또 당첨과 로또 자동 발행 로직 분리
- property에 bonusNumber 삭제

wondroid-world committed:
chore: controller package 생성

wondroid-world committed:
refactor: 모델내 상수 -> 공통 상수로 변경

wondroid-world committed:
refactor: 모델내 상수 -> 공통 상수로 변경

wondroid-world committed:
refactor: property -> method 변경

wondroid-world committed:
refactor: listOf -> setOf 타입 변경

wondroid-world committed:
refactor: 함수의 접근제어자 변경

wondroid-world committed:
refactor: 로또 test 함수명 변경

wondroid-world committed:
refactor: 로또 controller 네이밍 변경 & 구조 변경

wondroid-world committed:
refactor: inputView, outputView 변경

wondroid-world committed:
refactor: 상수 값 6 -> 0으로 변경

Gyuil-Hwnag commented on src/main/kotlin/lotto/contants/Constants.kt:
```diff
@@ -0,0 +1,9 @@
+
+object Constants {
+    const val LOTTO_AMOUNT = 1_000
+    const val MINIMUM_NUMBER = 1
```
이전 코멘트에서 말씀해주신것처럼 추후에 Constants에 여러가지 상수 값들이 추가될 것 같은데, 어떤 것을 담당하고 있는지를 명확히 하기위해서 어떤 네이밍을 지어줄 수 있을까요?

wondroid-world replied:
아! 두루! 두루의 말대로 다음 사람이 이해하기 쉬운 네이밍이면 좋겠다라는 생각에 LottoRuleConstants로 변경했습니다. 이렇게 네이밍을 지으니, 더 직관적이네요!

d45d121240dc979a38d74fd69b847015395dabcc

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -0,0 +1,25 @@
+    private val numbers: Set<LottoNumber> = generateLotto(),
+) {
+    init {
+        require(numbers.size == Constants.LOTTO_PICK_COUNT)
```
LottoTicket 객체의 가변인자를 사용해서 생성도 가능합니다
모찌가 보시기에는 두개는 어떤 장단점을 가지고 있을까요?
~~~kotlin
val lottoTicket = LottoTicket(
                setOf(
                    LottoNumber(1),
                    LottoNumber(2),
                    LottoNumber(3),
                    LottoNumber(4),
                    LottoNumber(5),
                    LottoNumber(5),
                ),
            )

val lottoTicket = LottoTicket(
                LottoNumber(1),
                LottoNumber(2),
                LottoNumber(3),
                LottoNumber(4),
                LottoNumber(5),
                LottoNumber(5),
            )
~~~

wondroid-world replied:
위에 꺼로 사용을 하면, 다른 개발자에 굳이 필요없는 set으로 데이터를 보관하고 있다라는 정보를 주는 거 같네요! 그래서 가변인자와 가짜 생성자를 만들어서 수정을 해봤어요!

```
package lotto.model

import lotto.contants.LottoRuleConstants.LOTTO_PICK_COUNT

class WinningLotto(
    private val winningNumbers: Set<LottoNumber>,
    private val bonusNumber: LottoNumber,
) {
    init {
        require(winningNumbers.size == LOTTO_PICK_COUNT.value) { ERROR_NUMBERS_COUNT }
        require(!winningNumbers.contains(bonusNumber)) { ERROR_LOTTO_NUMBERS_NOT_CONTAIN_BONUS_NUMBER }
    }

    fun getResult(lottoTickets: List<LottoTicket>): LottoResult {
        val result = lottoTickets.map { getRank(it) }
        return LottoResult(result)
    }

    fun getRank(lottoTicket: LottoTicket): Rank {
        val countOfMatch: Int = getCountOfMatch(lottoTicket)
        val matchBonus: Boolean = getMatchBonus(lottoTicket)
        return Rank.valueOf(countOfMatch, matchBonus)
    }

    private fun getCountOfMatch(lottoTicket: LottoTicket): Int = winningNumbers.intersect(lottoTicket.getNumbers()).size

    private fun getMatchBonus(lottoTicket: LottoTicket): Boolean = lottoTicket.getNumbers().contains(bonusNumber)

    companion object {
        private const val ERROR_NUMBERS_COUNT = "로또 번호의 개수는 6개입니다."
        private const val ERROR_LOTTO_NUMBERS_NOT_CONTAIN_BONUS_NUMBER = "로또 당첨 번호는 보너스 번호를 포함하지 말아야합니다."

        fun create(vararg lottoNumbers: LottoNumber, bonusNumber: LottoNumber): WinningLotto {
            return WinningLotto(lottoNumbers.toSet(), bonusNumber)
        }
    }
}


```

wondroid-world replied:
d5c22c10acbf6ce0783fe6f9967d2bed3718e6e0

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -0,0 +1,25 @@
+                .sorted()
+                .map { LottoNumber(it) }
+                .toSet()
+    }
```
[이전 코멘트](https://github.com/woowacourse/kotlin-lotto/pull/118#discussion_r1965108971)

이전 코멘트에 이어서 한번 얘기를 나눠볼까 합니다!
만약 범위가 다른 미국 / 중국 / 프랑스의 로또가 있다면, 해당 티켓들을 만들기 위해서는 나라별 LottoTicket 객체가 모두 만들어지게 될 것 같은데, 이를 어떻게 유연하게 바꿔볼 수 있을까요?🤔

wondroid-world replied:
음.. 저라면, 해당 최소값과 최대값을 파라미터로 받아볼거 같아요! 그래서, generateLotto() 함수에 파라미터로 min과 max 값을 받아봤습니다. 그런데, 이렇게 만드니, min, max값을 LottoTicket 객체를 만들 때 매번 넣어줘야하더라구요. 

이걸 보니, "로또 티켓 역할과 로또 번호 생성자의 역할을 분리해야하나?"라는 결과에 도달했는 데,  아직 코드적으로 구체적으로 생각이 나지 않습니다... 😢

Gyuil-Hwnag replied:
로또 티켓의 역활과 생성자의 역활을 분리해보는 것은 어떤가? 라는 생각에서 시작을 했었습니다!!
모찌의 말대로, 이렇게 된다면 매번 생성시에 최대 / 최소를 입력으로 필요로 하게됩니다
`로또 티켓 / 생성자의 역활 분리를 어떻게 할까` -> `인터페이스로 만드는 방법도 있겠네` -> `아니면 Factory 함수로만 만들어 둘까?`

위와 같은 생각의 흐름을 해보셨으면 해서 남겨둔 코멘트 였습니다
반드시 적용을 해야 하는 것은 아니고, 이러한 고민과정에서 얻는 것이 있을거라 생각이 들어서, 고민을 해보셨던 것만 해도 아주 좋습니다👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,5 @@
+package lotto.view
+
+object InputView {
+    fun getUserInput(): String = readln().trim()
+}
```
[이전 코멘트](https://github.com/woowacourse/kotlin-lotto/pull/118#discussion_r1964970775)
InputView로 네이밍이 되어 있다면, 추후 로또 게임외에 다른 게임이 생긴다면 InputView 라는 네이밍은 한번 고민해볼만 하겠군요!
(이 부분에 대한 모찌의 생각이 궁금합니다~
패키지또는 모듈로 구분을 할 수 있고, InputView라는 네이밍을 통일성으로 가져가고 싶다등)

wondroid-world replied:
저는 InputView의 역할을 아주 좁게 생각하여, 인풋만 받는 class라고 생각을 합니다. 각 게임이 생긴다면 각 게임에 맞는 UserInterface를 만들겠습니다!

Gyuil-Hwnag replied:
모찌의 의견 정말 좋네요ㅎㅎ👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -0,0 +1,28 @@
+
+import lotto.contants.Constants.LOTTO_PICK_COUNT
+
+class WinningLotto(
+    private val winningNumbers: Set<LottoNumber>,
+    private val bonusNumber: LottoNumber,
+) {
```
WinningLotto 객체를 잘 만들어주셨네요~👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -0,0 +1,28 @@
+    private val winningNumbers: Set<LottoNumber>,
+    private val bonusNumber: LottoNumber,
+) {
+    init {
+        require(winningNumbers.size == LOTTO_PICK_COUNT)
+        require(!winningNumbers.contains(bonusNumber))
+    }
```
해당 조건들이 실패하는 경우에 대한 문구도 정의를 해보는 것은 어떨까 라는 생각이 듭니다~💪
만약 실제 서비스를 운영한다고 했을 때, 발생하는 에러에 대한 문구들이 명확히 정해져 있다면 CS 대응 및 사용자들에게도 명확히 인지를 시키는 장점도 있을 것 같습니다

wondroid-world replied:
맞아요! 구현만 하다보니, 이런 세심한 부분을 놓치는 거 같네요! 감사합니다 두루!

6949eaecbf157d9e1810834f9686f2a1444ac134

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/InputValidator.kt:
```diff
@@ -0,0 +1,36 @@
+import lotto.contants.Constants
+
+class InputValidator {
+    fun validateAmount(amount: String) {
+        require(amount.toIntOrNull() != null) {
+            IllegalArgumentException(ERROR_INVALID_AMOUNT)
+        }
+        require(amount.toInt() > Constants.ZERO) {
+            IllegalArgumentException(ERROR_INVALID_OVER_ZERO)
+        }
+        require(amount.toInt() % Constants.LOTTO_AMOUNT == Constants.ZERO) {
+            IllegalArgumentException(ERROR_INVALID_DIVISION)
+        }
+    }
```
하나의 함수에서 너무 많은 기능을 하는 것 같다는 생각이 드는데, 함수의 기능을 분리해보는 것은 어떨까 라는 생각이 듭니다!
그렇다면, Input 에 여러가지 기능들이 추가되었을 때도 활용할 수 있겠군요

wondroid-world replied:
두루, 
함수는 한 역할만 한다라는 사실을 까먹고 있었네요! 두루말처럼 한 함수가 한 역할만 하도록 구현을 하니, 제가 구입 금액을 받는 부분과 보너스 번호 검증하는 부분이 똑같았지만, 2번 구현하고 있었네요!

두루덕분에 더 깔끔하게 코드를 수정할 수 있었습니다! 감사해요 두루!

wondroid-world replied:
5a4d1ae07b0513f452fd653889e9c77ac3053d5b

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/InputValidator.kt:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    companion object {
+        private const val ERROR_INVALID_AMOUNT = "구입금액은 정수를 입력해야 합니다."
+        private const val ERROR_INVALID_OVER_ZERO = "구입금액은 0보다 커야 합니다."
+        private const val ERROR_INVALID_DIVISION = "구입금액은 1000의 배수여야 합니다."
+        private const val ERROR_INVALID_WINNING_TYPE = "당첨번호는 정수를 입력해야 합니다."
+    }
```
에러 문구에 대한 정의 너무 잘해주셨네요~👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/UserInterface.kt:
```diff
@@ -0,0 +1,72 @@
+import lotto.model.Rank
+import java.text.DecimalFormat
+
+class UserInterface(
+    private val inputValidator: InputValidator = InputValidator(),
+) {
```
UserInterface 라는 것을 만들어주셨군요!
UserInterface가 만들어지면서, 뭔가 InputView / OutputView는 println()만 있을뿐 존재가 하는 의미가 없다는 느낌이 들기는 하네요
UserInterface를 만들게 된 이유가 궁금합니다~
InputView / OutputView / UserInterface 간의 명확한 역활이 구분되었다면, 더 좋은 구조가 되었을 것 같다는 생각이 듭니다💪

wondroid-world replied:
저의 생각에서 
InputView -> 입력을 받는 역할
OuputView -> 출력을 하는 역할
이렇게 두 클래스가 서로의 역할을 잘 해줘야한다고 생각했습니다.

예로, 구입금액을 입력받기 위해서 inputView에서 "구입금액을 입력해 주세요."를 출력하는 게 각 클래스의 역할에 맞지 않는다고 생각이들었습니다. 그래서 inputView와 outputView를 중재하는 역할이 필요해서 UserInterface가 만들어졌습니다.

Gyuil-Hwnag replied:
그렇군요!!
input / output View에서 담당하는 역활이 너무 없어 보여서, 아예 없어지고 LottoView만 있어도 될 것 처럼 보였습니다
근데, 이렇게 하다보면 `입력 / 출력`은 서로 역활이 다른데, 한곳에서 같이 처리를 하는게 맞을까? 라는 생각이 들긴 하네요~💪
(Input / Output이 없어지고 UserView만 존재 or input / output의 역활 분배)

Gyuil-Hwnag reviewed (changes requested):
모찌 안녕하세요!

해당 미션 리뷰 반영 고생 많으셨습니다!
리뷰 반영하면서 구현한 내용 수정 너무 잘 해주셨습니다 💯

이번에 코멘트를 반영하면서, 구조가 변경된 걸 보는데 페어분하고 고민을 많이 하셨던게 느껴졌네요!
다음 미션 진행전에 모찌가 생각해보시기에 개선이 필요한 내용들이 있다면 변경하거나 코멘트 남겨서 같이 PR요청주세요

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

wondroid-world committed:
refactor: 상수 네이밍 Constant -> LottoRuleConstant로 변경

wondroid-world committed:
refactor : validate 함수가 한개의 하나의 기능을 하도록 변경

wondroid-world committed:
fix : LottoRuleConstant 값으로 변경

wondroid-world committed:
refactor: LottoNumber, LottoTicket, WinningLotto애 대한 에러 메시지 작성

wondroid-world committed:
refactor: 가독성 있도록 코드 변경

wondroid-world committed:
refactor: 가변인자와 가짜 생성자 함수 사용

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/LottoNumber.kt:
```diff
@@ -0,0 +1,15 @@
+data class LottoNumber(
+    val number: Int,
+) {
+    init {
+        require(number in LottoRuleConstants.MINIMUM_NUMBER.value..LottoRuleConstants.MAXIMUM_NUMBER.value) { ERROR_INVALID_RANGE }
+    }
```
에러 문구에 대해서 추가를 해주셨군요!👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/view/InputValidator.kt:
```diff
@@ -0,0 +1,22 @@
+import lotto.contants.LottoRuleConstants
+
+class InputValidator {
+    fun validateInteger(input: String) = require(input.toIntOrNull() != null) { ERROR_INVALID_INTEGER }
+
+    fun validateOverZero(input: String) = require(input.toInt() > LottoRuleConstants.ZERO.value) { ERROR_INVALID_OVER_ZERO }
+
+    fun validateAmountUnits1000(input: String) =
+        require(
+            input.toInt() % LottoRuleConstants.LOTTO_AMOUNT.value == LottoRuleConstants.ZERO.value,
+        ) {
+            ERROR_INVALID_DIVISION
+        }
```
각각의 로직들에 대해서 재사용이 가능해졌군요!👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -0,0 +1,37 @@
+        private const val ERROR_NUMBERS_COUNT = "로또 번호의 개수는 6개입니다."
+        private const val ERROR_LOTTO_NUMBERS_NOT_CONTAIN_BONUS_NUMBER = "로또 당첨 번호는 보너스 번호를 포함하지 말아야합니다."
+
+        fun create(vararg lottoNumbers: LottoNumber, bonusNumber: LottoNumber): WinningLotto {
+            return WinningLotto(lottoNumbers.toSet(), bonusNumber)
+        }
```
가변인자를 받는 Fake 생성자를 만들어 주셨군요~👍

Gyuil-Hwnag commented on src/main/kotlin/lotto/contants/LottoRuleConstants.kt:
```diff
@@ -0,0 +1,11 @@
+package lotto.contants
+
+enum class LottoRuleConstants(
+    val value: Int,
+) {
+    LOTTO_AMOUNT(1_000),
+    MINIMUM_NUMBER(1),
+    MAXIMUM_NUMBER(45),
+    LOTTO_PICK_COUNT(6),
+    ZERO(0),
+}
```
해당 내용을 enum 객체로 만들어준 이유가 있을까요??🤔

Gyuil-Hwnag reviewed (approved):
모찌 안녕하세요~
해당 미션 너무 고생 많으셨습니다!

리뷰 반영 정말 잘 해주셨네요💯
해당 미션은 다음 미션 진행을 위하여 머지하도록 하겠습니다.

도움이 될만한 코멘트 달았으니 확인하여 다음 미션에서 반영해주셔도 좋을 것 같습니다.
`UserView라는 `input / output`이 한곳에 모여있는 구조`에 대해서는 다음미션을 진행하면서, 더 나눠볼 얘기가 있을 것 같아서 다음미션을 진행하면서 한번 고민을 해봐도 좋을 것 같습니다!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪
